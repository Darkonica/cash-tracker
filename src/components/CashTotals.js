import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectOperations from '../selectors/operations';

export const CashTotals = (props) => {
	const expensesAmount = props.operations
		.filter((expense) => {
			return expense.selectedSwitchState === 'expense'
		})
		.map((expense) => (expense.amount))
		.reduce((sum, value) => sum + value, 0);

	const incomesAmount = props.operations
		.filter((income) => {
			return income.selectedSwitchState === 'income'
		})
		.map((income) => (income.amount))
		.reduce((sum, value) => sum + value, 0);

	return (
		<div className="cash-totals pure-u-md-1-2">
			<div className="block pure-u-md-1-3">
				<h3>Balance</h3>
				<div className="amount">{numeral((incomesAmount - expensesAmount) / 100).format('$0.00')}</div>
			</div>
			<div className="block pure-u-md-1-3">
				<h3>Incomes</h3>
				<div className="amount">{numeral(incomesAmount / 100).format('$0.00')}</div>
			</div>
			<div className="block pure-u-md-1-3">
				<h3>Expenses</h3>
				<div className="amount">{numeral(expensesAmount / 100).format('$0.00')}</div>
			</div>
		</div>
	);
};

const mapStateToProps = (state) => {
	return {
		operations: selectOperations(state.operations, state.filters)
	};
};

export default connect(mapStateToProps)(CashTotals);