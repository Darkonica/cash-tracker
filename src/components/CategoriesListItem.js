import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { removeCategory } from '../actions/categories';

class CategoriesListItem extends React.Component {
	constructor(props) {
		super(props);
	};
	onRemove = () => {
		this.props.removeCategory({ id: this.props.category.id, categoryPart: this.props.categoryPart });
	};
	render() {
		return (
			<div className="item-cont">

				<div className="info">
					{this.props.category.name}
				</div>

				<div className="actions">
					<div
						className="remove"
						onClick={this.onRemove}
					>remove</div>
					<div className="edit">edit</div>
				</div>
				
			</div>
		);
	}
};

const mapDispatchToProps = (dispatch, props) => ({
	removeCategory: (data) => dispatch(removeCategory(data)),
});

export default connect(null, mapDispatchToProps)(CategoriesListItem);