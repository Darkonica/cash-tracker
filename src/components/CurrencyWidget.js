import React from 'react';
import { parseString } from 'xml2js';

export class CurrencyWidget extends React.Component {
		constructor(props) {
			super(props);

			this.state = {
				data: {},
				isLoading: false
			};
		};

		getData(){
      return fetch('https://api.fixer.io/latest?base=USD')
	      .then((response) => response.json())
	      .then(data => this.setState({ data: data.rates, isLoading: false }))
	      .catch((error) => {
	        console.error(error);
	      });
    }

    componentDidMount() {
    	this.setState({ isLoading: true });
      this.getData();
    };
		
		render() {
			return (
				<div className="currency-widget">
					<h1>Exchange rates</h1>

					<div className="flex-cont pure-u-sm-1-2">
						<div className="block">
							<div className="icon"></div>
						</div>
						<div className="block">
							<h3>EUR</h3>
							<span className="rate">{this.state.data.EUR && this.state.data.EUR}$</span>
						</div>
						<div className="block">
							<h3>GBP</h3>
							<span className="rate">{this.state.data.GBP && this.state.data.GBP}$</span>
						</div>
					</div>
					
				</div>
			);
		}
};

export default CurrencyWidget;