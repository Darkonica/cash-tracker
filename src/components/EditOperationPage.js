import React from 'react';
import { connect } from 'react-redux';
import OperationForm from './OperationForm';
import { editOperation, removeOperation } from '../actions/operations';

export class EditOperationPage extends React.Component {
	onSubmit = (operation) => {
		this.props.editOperation(this.props.operation.id, operation);
		this.props.history.push('/');
	};
	onRemove = () => {
		this.props.removeOperation({ id: this.props.operation.id });
		this.props.history.push('/');
	};
	render() {
		return (
			<div>
				<OperationForm
					expense={this.props.operation}
					onSubmit={this.onSubmit}
				/>
				<button onClick={this.onRemove}>Remove</button>
			</div>
		);
	}
	
};

const mapStateToProps = (state, props) => {
	return {
		operation: state.operations.find((operation) => {
			return operation.id === props.match.params.id;
		})
	};
};

const mapDispatchToProps = (dispatch, props) => ({
	editOperation: (id, operation) => dispatch(editOperation(id, operation)),
	removeOperation: (data) => dispatch(removeOperation(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditOperationPage);