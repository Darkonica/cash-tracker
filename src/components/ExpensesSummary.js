import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = (props) => {
	const expenseCount = props.expenses.length;
	const expensesTotal = props.expenses
		.map((expense) => (expense.amount))
		.reduce((sum, value) => sum + value, 0);

	return (
		<div>
			<p>Viewing {expenseCount === 1 ? `1 expense` : `${expenseCount} expenses`} totalling {numeral(expensesTotal / 100).format('$0.00')}</p>
		</div>
	);
};

const mapStateToProps = (state) => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	};
};

export default connect(mapStateToProps)(ExpensesSummary);