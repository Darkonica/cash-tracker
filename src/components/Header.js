import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
	<header>
		<div className="logo-text">Cash Tracker</div>
		<nav>
			<ul>
				<li><NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink></li>
				<li><NavLink to="/settings" activeClassName="is-active">Settings</NavLink></li>
			</ul>
		</nav>
	</header>
);

export default Header;