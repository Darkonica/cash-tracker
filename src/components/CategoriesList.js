import React from 'react';
import { connect } from 'react-redux';
import CategoriesListItem from './CategoriesListItem';
import { addCategory } from '../actions/categories';

class CategoriesList extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			name: '',
			error: ''
		};
	};

	onNameChange = (e) => {
		const name = e.target.value;
		this.setState(() => ({ name }));
	};
	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.name) {
			console.log('error');
			this.setState(() => ({ error: 'Please, set some values' }));
		} else {
			this.setState(() => ({ error: '' }))
			this.props.addCategory({
				name: this.state.name,
				categoryPart: this.props.categoryPart
			});
		}
	};

	render() {
		return (
			<div>
				{
					this.props.categories[this.props.categoryPart].length === 0 ? (
						<p>No categories</p>
					) : (
						this.props.categories[this.props.categoryPart].map((category) => {
							return <CategoriesListItem key={category.id} category={category} categoryPart={this.props.categoryPart} />
						})
					)
				}
				<form
					className="add-category-form"
					onSubmit={this.onSubmit}
				>
					<div className="block pure-u-md-2-3">
						<input
							type="text"
							placeholder="Name"
							value={this.state.name}
							onChange={this.onNameChange}
						/>
					</div>
					<div className="pure-u-md-1-3">
						<button className="btn-active">Add category</button>
					</div>
				</form>
			</div>
		);
	}
};

const mapStateToProps = (state) => {
	return {
		categories: state.categories
	};
};

const mapDispatchToProps = (dispatch, props) => ({
	addCategory: (category) => dispatch(addCategory(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(CategoriesList);