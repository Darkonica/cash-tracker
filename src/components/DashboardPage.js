import React from 'react';
import { connect } from 'react-redux';
import OperationList from './OperationList';
import OperationForm from './OperationForm';
import CashTotals from './CashTotals';
import { addOperation } from '../actions/operations';
import CurrencyWidget from './CurrencyWidget';

export class DashboardPage extends React.Component {
	onSubmit = (operation) => {
		this.props.addOperation(operation);
	};
	render() {
		return (
			<div className="dashboard">
				<OperationForm
					categories={this.props.categories}
					onSubmit={this.onSubmit}
				/>
				<CashTotals />
				<div className="pure-u-md-1-2">
					<CurrencyWidget />
				</div>
				<OperationList />
			</div>
		);
	}
};



const mapStateToProps = (state, props) => {
	return {
		categories: state.categories
	};
};

const mapDispatchToProps = (dispatch, props) => ({
	addOperation: (operation) => dispatch(addOperation(operation))
});

export default connect(mapStateToProps, mapDispatchToProps)(DashboardPage);