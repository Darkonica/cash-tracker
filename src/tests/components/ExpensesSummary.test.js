import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('render ExpensesSummary', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[]}/>)
	expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary with 1 expense', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[expenses[1]]} />)
	expect(wrapper).toMatchSnapshot();
});