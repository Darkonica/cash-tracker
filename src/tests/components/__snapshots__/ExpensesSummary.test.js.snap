// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpensesSummary 1`] = `
<div>
  <p>
    Viewing 
    0 expenses
     totalling 
    $0.00
  </p>
</div>
`;

exports[`render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1 expense
     totalling 
    $1095.00
  </p>
</div>
`;
